<!-- Dynamically create individual sample rmd pages under the same datanav menu -->
`r paste0('# ', sample, '{data-navmenu="Individual Samples"}')`

```{r, sample_setup, include=FALSE, echo=FALSE}
cov_f <- paste0("./base_coverages/", sample, ".per_base_coverage.bed")
run_depth_plot <- FALSE
if (file.exists(cov_f)) {
  run_depth_plot <- TRUE
}
```

Row
------------------------------

### Sample Name
```{r, echo=FALSE, error=TRUE}
valueBox(sample, icon = "")
```

### Genome Completeness
```{r, echo=FALSE, error=TRUE}
completeness <- sample_df$genome_completeness[1]
sec_color <- get_genome_completeness_color(completeness)
valueBox(
    completeness,
    icon = "",
    color = sec_color
)
```

### Sample Status
```{r, echo=FALSE, error=TRUE}
# Based on run satus, set colour and icon
sample_status <- sample_df$qc_pass[1]
if ( sample_status == 'PASS' ) {
    sec_icon = ""
    sec_color = "success"
} else if ( grepl(sample_status, 'POTENTIAL_FRAMESHIFTS') ) {
    sec_icon = ""
    sec_color = "orange"
} else if ( grepl(sample_status, 'PARTIAL_GENOME') ) {
    sec_icon = ""
    sec_color = "warning"
} else {
    sec_icon = ""
    sec_color = "danger"
}
valueBox(
  sample_status,
  icon = sec_icon,
  color = sec_color
)
```

Row {data-height=600}
------------------------------

### {-}

```{r, variant_plot, echo=FALSE, error=TRUE}
# Create table from bam file csv showing variants and subvariants
var_f <- paste0("./all_variation_positions/", sample, "_variation.csv")
if (file.exists(var_f)) {
  variants <- read.table(var_f, header = TRUE, sep = ",")
} else {
    variants <- data.frame(matrix(ncol = 12, nrow = 0))
    colv <- c("position", "ref_base", "percentage_nonref", "variant_type", "total_reads", "A", "T", "C", "G", "del", "ins", "distribution")
    colnames(variants) <- colv
  }

  # Text info for hover
  text_info <- paste(
    'Position: ', variants[['position']], '<br>',
    'Reference: ', variants[['ref_base']], '<br>',
    'Total Reads: ', variants[['total_reads']], '<br>',
    '  A: ', variants[['A']], '<br>',
    '  T: ', variants[['T']], '<br>',
    '  G: ', variants[['G']], '<br>',
    '  C: ', variants[['C']], '<br>',
    '  Del:', variants[['del']], '<br>',
    '  Ins:', variants[['ins']], '<br>',
    'Insertion Distribution: ', variants[['distribution']], '<br>',
    sep = ''
  )
  variants$text_info <- text_info

  # Plot variation table
  #  Also customdata instead of text due to https://github.com/plotly/plotly.R/issues/1859
  p <- plot_ly(variants) %>%
    add_trace(x = variants[['position']], y = variants[['percentage_nonref']],
        type = 'scatter', mode = 'markers',
        color = ~factor(variant_type, levels = c("A SNP", "T SNP", "G SNP", "C SNP", "Ref", "Mixed", "Deletion", "Insertion")),
        marker = list(sizemode = 'diameter', size = 15.0, opacity = 0.7),
        customdata = ~text_info,
        hovertemplate = paste(
          '<b>Percentage</b>: %{y:.2f}<br>',
          '%{customdata}<extra></extra>',
          sep=''
        )
    )

  # Add final layout to plot
  p <- p %>%
    layout(
      xaxis = list (title = "Position", tickformat = ".0f"),
      yaxis = list (title = "Percent Non Reference Base", tickformat = '2.%'),
      shapes = hline(75)
    )
p
```

> **Figure 1.** Variant Read Counts. The sample pileup was analyzed to find positions in the genome where
> the alternate allele percentage was greater-than or equal to 15%.

### **Table 1.** Consensus Called Variants {data-width=320}
```{r, variant_table, echo=FALSE, error=TRUE}
# Create table of called variants using the vcf file turned into a tsv
var_f <- paste0("./variant_tsvs/", sample, ".vcf.tsv")
if (file.exists(var_f)) {
  variants <- read.table(var_f, header = TRUE, sep = "\t")
  ddt <- variants %>%
    DT::datatable(rownames = FALSE,
      options = list(
        scrollX = FALSE
    ))
} else {
  variants <- data.frame(matrix(ncol = 4, nrow = 0))
  colv <- c("Pos", "Ref", "Alt", "Qual")
  colnames(variants) <- colv
  ddt <- variants %>%
    DT::datatable(rownames = FALSE)
}
ddt
```

Row
------------------------------
### {-}
```{r, stat_cov_plot, echo=FALSE, error=TRUE, fig.width=12, fig.height=3, eval=run_depth_plot}
# Read in data and select columns
cov_f <- paste0("./base_coverages/", sample, ".per_base_coverage.bed")
coverage <- read.table(cov_f, header = TRUE, sep = "\t")
coverage <- select(coverage, position, depth)

# Create colour ranges for the depth plot
#  For if nothing is above 20x coverage
if (max(coverage$depth) < 20) {
  rect_data <- data.frame(xmin=min(coverage$position),
                          xmax=max(coverage$position),
                          ymin=c(0),
                          ymax=c(20),
                          col=c("red"))
# For if nothing is above 100x coverage
} else if (max(coverage$depth) <= 100) {
  rect_data <- data.frame(xmin=min(coverage$position),
                          xmax=max(coverage$position),
                          ymin=c(0,20),
                          ymax=c(20,100),
                          col=c("red","yellow"))
} else {
  rect_data <- data.frame(xmin=min(coverage$position),
                          xmax=max(coverage$position),
                          ymin=c(0,20,100),
                          ymax=c(20,100,max(coverage$depth)),
                          col=c("red","yellow","green"))
}

# Create Plot
ggplot(coverage) +
  geom_line(aes(x=position, y=depth)) +
  scale_y_log10() +
  xlab("Genome Position (nt)") +
  ylab("Sequencing Depth") +
  geom_rect(data=rect_data, aes(xmin=xmin,xmax=xmax,ymin=ymin,ymax=ymax,fill=col),alpha=0.1) +
  scale_fill_identity()
```

> **Figure 2.** Genome Coverage Plot. Sequencing depth at each position in the genome
